File: Medium/C++/sum-of-two-integers.cpp

# include <iostream>
/*
int getSum(int a, int b){
	// std::cout << a << ' ' << b << std::endl;
	while (b){
		unsigned int carry = (a & b) << 1 ;
		a = a ^ b;
		b = carry;
	}
	return a;
}
*/

int getSum(int a, int b) {
    while (b){
        int carry = (a&b);
        a = a^b;
        b = (unsigned int)carry << 1;
    }
    return a;
}


int main(){
	std::cout << getSum(32, -17);
	return 0;
}

File: Python/Binary/sum-of-two-integers.py

def getSum(a:int, b:int) -> int:
	mask = (1<<32)-1
	while b&mask:
		carry = (a&b) << 1 # all this gets is the carry value and shifts it to the left
		a = a ^ b  # this adds them, but doesn't account for the carry 1
		b = carry

	return a & mask if b > 0 else a

print(getSum(-11, -9))

# 0b11111111111111111111111111110110